File: core/src/main/java/com/tunjid/androidbootstrap/core/view/ViewHider.java
package com.tunjid.androidbootstrap.core.view;


import android.graphics.Point;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.core.view.ViewCompat;
import androidx.core.view.ViewPropertyAnimatorCompat;
import androidx.core.view.ViewPropertyAnimatorListener;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.WindowManager;
import android.view.animation.Interpolator;

import java.lang.annotation.Retention;
import java.util.ArrayList;
import java.util.List;

import static android.content.Context.WINDOW_SERVICE;
import static java.lang.annotation.RetentionPolicy.SOURCE;

/**
 * Translates a view offscreen, useful for adding the quick return pattern to a view.
 * <p>
 * Created by tj.dahunsi on 2/19/16.
 */
public class ViewHider {

    private static final Interpolator FAST_OUT_SLOW_IN_INTERPOLATOR = new FastOutSlowInInterpolator();

    @Retention(SOURCE)
    @IntDef({LEFT, TOP, RIGHT, BOTTOM})
    @SuppressWarnings("WeakerAccess")
    public @interface HideDirection {
    }

    public static final int LEFT = 0;
    public static final int TOP = 1;
    public static final int RIGHT = 2;
    public static final int BOTTOM = 3;

    private boolean isVisible = true;
    private final @HideDirection int direction;
    private final long duration;

    private final View view;
    private final Listener listener;

    private final Runnable startRunnable;
    private final Runnable endRunnable;

    public static Builder of(View view) {return new Builder(view);}

    private ViewHider(View view, Listener listener, @HideDirection int direction, long duration) {
        this.view = view;
        this.listener = listener;
        this.duration = duration;
        this.direction = direction;

        startRunnable = () -> {if (isVisible) view.setVisibility(View.VISIBLE);};
        endRunnable = () -> {if (!isVisible) view.setVisibility(View.GONE);};
    }

    public void show() {
        toggle(true);
    }

    public void hide() {
        toggle(false);
    }

    private void toggle(final boolean visible) {
        if (this.isVisible != visible) {
            final ViewTreeObserver original = view.getViewTreeObserver();

            // View hasn't been laid out yet and has it's observer attached
            if (view.getHeight() == 0 && original.isAlive()) {
                original.addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
                    @Override
                    public boolean onPreDraw() {
                        ViewTreeObserver current = view.getViewTreeObserver();
                        if (current.isAlive()) current.removeOnPreDrawListener(this);
                        // toggle as soon as the view is visible
                        toggle(visible);
                        return true;
                    }
                });
                return;
            }

            this.isVisible = visible;
            float displacement = visible ? 0 : getDistanceOffscreen();

            ViewPropertyAnimatorCompat animator = ViewCompat.animate(view)
                    .setDuration(duration)
                    .setInterpolator(FAST_OUT_SLOW_IN_INTERPOLATOR)
                    .setListener(listener);

            if (direction == LEFT || direction == RIGHT) animator.translationX(displacement);
            else animator.translationY(displacement);

            animator.start();
        }
    }

    private int getDistanceOffscreen() {
        int[] location = new int[2];
        Point displaySize = new Point();

        WindowManager manager = (WindowManager) view.getContext().getSystemService(WINDOW_SERVICE);

        view.getLocationInWindow(location);
        manager.getDefaultDisplay().getSize(displaySize);

        // These calculations don't take the status bar into account, unlikely to matter however
        switch (direction) {
            case LEFT:
                return -(location[0] + view.getWidth());
            case TOP:
                return -(location[1] + view.getHeight());
            case RIGHT:
                return displaySize.x - location[0];
            case BOTTOM:
                return displaySize.y - location[1];
            default:
                throw new IllegalArgumentException("Invalid direction");
        }
    }

    public static class Builder {

        @HideDirection
        private int direction = BOTTOM;
        private long duration = 200L;
        private final View view;
        private final Listener listener = new Listener();

        Builder(@NonNull View view) {this.view = view;}

        public Builder setDirection(@HideDirection int direction) {
            this.direction = direction;
            return this;
        }

        @SuppressWarnings("unused")
        public Builder setDuration(long duration) {
            this.duration = duration;
            return this;
        }

        @SuppressWarnings("unused")
        public Builder addStartRunnable(@NonNull Runnable runnable) {
            listener.startRunnables.add(runnable);
            return this;
        }

        @SuppressWarnings("unused")
        public Builder addEndRunnable(@NonNull Runnable runnable) {
            listener.endRunnables.add(runnable);
            return this;
        }

        public ViewHider build() {
            ViewHider viewHider = new ViewHider(view, listener, direction, duration);
            listener.startRunnables.add(0, viewHider.startRunnable);
            listener.endRunnables.add(0, viewHider.endRunnable);

            return viewHider;
        }
    }

    private static class Listener implements ViewPropertyAnimatorListener {

        private final List<Runnable> startRunnables = new ArrayList<>();
        private final List<Runnable> endRunnables = new ArrayList<>();

        Listener() {}

        @Override
        public void onAnimationStart(View view) {
            if (!startRunnables.isEmpty()) for (Runnable runnable : startRunnables) runnable.run();
        }

        @Override
        public void onAnimationEnd(View view) {
            if (!endRunnables.isEmpty()) for (Runnable runnable : endRunnables) runnable.run();
        }

        @Override
        public void onAnimationCancel(View view) {}
    }
}

File: view/src/main/java/com/tunjid/androidbootstrap/view/animator/ViewHider.kt
package com.tunjid.androidbootstrap.view.animator


import android.content.Context.WINDOW_SERVICE
import android.graphics.Point
import android.view.View
import android.view.WindowManager
import androidx.annotation.IntDef
import androidx.core.view.doOnPreDraw
import androidx.dynamicanimation.animation.DynamicAnimation
import androidx.dynamicanimation.animation.SpringAnimation
import com.tunjid.androidbootstrap.view.util.spring

/**
 * Translates a view offscreen, useful for adding the quick return pattern to a view.
 *
 *
 * Created by tj.dahunsi on 2/19/16.
 */
class ViewHider private constructor(
        private val view: View,
        private val listener: Listener,
        @param:HideDirection @field:HideDirection
        private val direction: Int,
        private val options: SpringAnimation.() -> Unit) {

    private var isVisible = true

    private val startRunnable: () -> Unit = { if (isVisible) view.visibility = View.VISIBLE }
    private val endRunnable: () -> Unit = { if (!isVisible) view.visibility = View.GONE }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(LEFT, TOP, RIGHT, BOTTOM)
    annotation class HideDirection

    fun show() = toggle(true)

    fun hide() = toggle(false)

    private fun toggle(visible: Boolean) {
        if (this.isVisible == visible) return

        // View hasn't been laid out yet and has it's observer attached
        if (!view.isLaidOut) return view.doOnPreDraw { toggle(visible) }

        this.isVisible = visible
        val displacement = (if (visible) 0 else getDistanceOffscreen()).toFloat()

        view.spring(
                if (direction == LEFT || direction == RIGHT) SpringAnimation.TRANSLATION_X
                else SpringAnimation.TRANSLATION_Y
        )
                .apply {
                    options.invoke(this)
                    for (runnable in listener.startRunnables) runnable.invoke()
                }
                .withOneShotEndListener { for (runnable in listener.endRunnables) runnable.invoke() }
                .animateToFinalPosition(displacement)
    }

    // These calculations don't take the status bar into account, unlikely to matter however
    private fun getDistanceOffscreen(): Int {
        val location = IntArray(2)
        val displaySize = Point()

        val manager = view.context.getSystemService(WINDOW_SERVICE) as? WindowManager
                ?: return 0

        view.getLocationInWindow(location)
        manager.defaultDisplay.getSize(displaySize)

        return when (direction) {
            LEFT -> -(location[0] + view.width)
            TOP -> -(location[1] + view.height)
            RIGHT -> displaySize.x - location[0]
            BOTTOM -> displaySize.y - location[1]
            else -> throw IllegalArgumentException("Invalid direction")
        }
    }

    class Builder internal constructor(private val view: View) {

        @HideDirection
        private var direction = BOTTOM
        private var options: SpringAnimation.() -> Unit = {}
        private val listener = Listener()

        fun setDirection(@HideDirection direction: Int): Builder = apply { this.direction = direction }

        fun addOptions(options: SpringAnimation.() -> Unit): Builder = apply { this.options = options }

        fun addStartRunnable(runnable: () -> Unit): Builder = apply { listener.startRunnables.add(runnable) }

        fun addEndRunnable(runnable: () -> Unit): Builder = apply { listener.endRunnables.add(runnable) }

        fun build(): ViewHider = ViewHider(view, listener, direction, options).apply {
            listener.startRunnables.add(0, startRunnable)
            listener.endRunnables.add(0, endRunnable)
        }
    }

    private class Listener internal constructor() {
        val startRunnables = mutableListOf<() -> Unit>()
        val endRunnables = mutableListOf<() -> Unit>()
    }

    private fun SpringAnimation.withOneShotEndListener(onEnd: (canceled: Boolean) -> Unit) = apply {
        addEndListener(object : DynamicAnimation.OnAnimationEndListener {
            override fun onAnimationEnd(animation: DynamicAnimation<out DynamicAnimation<*>>?, canceled: Boolean, value: Float, velocity: Float) {
                removeEndListener(this)
                onEnd(canceled)
            }
        })
    }

    companion object {

        const val LEFT = 0
        const val TOP = 1
        const val RIGHT = 2
        const val BOTTOM = 3

        fun of(view: View): Builder = Builder(view)
    }
}